src/stream/sosemanuk/sosemanuk.c

```diff
@@ -0,0 +1,816 @@
+        exit(EXIT_FAILURE);
+    }
+    XMEMCPY(wbuf, key, keylen);
+    if (keylen < 32) {
+        wbuf[keylen] = 0x01;
+        if (keylen < 31)
+            XMEMSET(wbuf + keylen + 1, 0, 31 - keylen);
+    }
+
+    LOAD32L(w0, wbuf);
+    LOAD32L(w1, wbuf + 4);
+    LOAD32L(w2, wbuf + 8);
+    LOAD32L(w3, wbuf + 12);
+    LOAD32L(w4, wbuf + 16);
+    LOAD32L(w5, wbuf + 20);
+    LOAD32L(w6, wbuf + 24);
+    LOAD32L(w7, wbuf + 28);
+
+    WUP0(0);   SKS3;
+    WUP1(4);   SKS2;
+    WUP0(8);   SKS1;
+    WUP1(12);  SKS0;
+    WUP0(16);  SKS7;
+    WUP1(20);  SKS6;
+    WUP0(24);  SKS5;
+    WUP1(28);  SKS4;
+    WUP0(32);  SKS3;
+    WUP1(36);  SKS2;
+    WUP0(40);  SKS1;
+    WUP1(44);  SKS0;
+    WUP0(48);  SKS7;
+    WUP1(52);  SKS6;
+    WUP0(56);  SKS5;
+    WUP1(60);  SKS4;
+    WUP0(64);  SKS3;
+    WUP1(68);  SKS2;
+    WUP0(72);  SKS1;
+    WUP1(76);  SKS0;
+    WUP0(80);  SKS7;
+    WUP1(84);  SKS6;
+    WUP0(88);  SKS5;
+    WUP1(92);  SKS4;
+    WUP0(96);  SKS3;
+
+#undef SKS
+#undef SKS0
+#undef SKS1
+#undef SKS2
+#undef SKS3
+#undef SKS4
+#undef SKS5
+#undef SKS6
+#undef SKS7
+#undef WUP
+#undef WUP0
+#undef WUP1
+
+    return CRYPT_OK;
+}
+
+
+/*
+ * Cipher initialization: the cipher internal state is initialized, using
+ * the provided key context and IV. The IV length is up to 16 bytes. If
+ * "ivlen" is 0 (no IV), then the "iv" parameter can be NULL.
+ * @param ss       The Sosemanuk state
+ * @param iv       Initialization vector
+ * @param ivlen    Length of iv
+ * @return CRYPT_OK on success
+ */
+int sosemanuk_setiv(sosemanuk_state *ss, unsigned char *iv, unsigned long ivlen)
+{
+
+    /*
+     * The Serpent key addition step.
+     */
+#define KA(zc, x0, x1, x2, x3)  do { \
+        x0 ^= ss->kc[(zc)]; \
+        x1 ^= ss->kc[(zc) + 1]; \
+        x2 ^= ss->kc[(zc) + 2]; \
+        x3 ^= ss->kc[(zc) + 3]; \
+    } while (0)
+
+    /*
+     * One Serpent round.
+     *   zc = current subkey counter
+     *   S = S-box macro for this round
+     *   i0 to i4 = input register numbers (the fifth is a scratch register)
+     *   o0 to o3 = output register numbers
+     */
+#define FSS(zc, S, i0, i1, i2, i3, i4, o0, o1, o2, o3)  do { \
+        KA(zc, r ## i0, r ## i1, r ## i2, r ## i3); \
+        S(r ## i0, r ## i1, r ## i2, r ## i3, r ## i4); \
+        SERPENT_LT(r ## o0, r ## o1, r ## o2, r ## o3); \
+    } while (0)
+
+    /*
+     * Last Serpent round. Contrary to the "true" Serpent, we keep
+     * the linear transformation for that last round.
+     */
+#define FSF(zc, S, i0, i1, i2, i3, i4, o0, o1, o2, o3)  do { \
+        KA(zc, r ## i0, r ## i1, r ## i2, r ## i3); \
+        S(r ## i0, r ## i1, r ## i2, r ## i3, r ## i4); \
+        SERPENT_LT(r ## o0, r ## o1, r ## o2, r ## o3); \
+        KA(zc + 4, r ## o0, r ## o1, r ## o2, r ## o3); \
+    } while (0)
+
+    ulong32 r0, r1, r2, r3, r4;
+    unsigned char ivtmp[16] = {0};
+
+    LTC_ARGCHK(ss != NULL);
+    LTC_ARGCHK(iv != NULL);
```

Comment:
If zero-length IV is acceptable we should perhaps also accept: `sosemanuk_setiv(&state, NULL, 0)` so perhaps something like `LTC_ARGCHK(iv != NULL || ivlen ==0);`